### PR TITLE
Revert the unintended change in tests reordering from #11220

### LINKED ISSUE #12008: pytest 8.0 sorting tests with multiple parameterization is broken
Сontinue #11976

The solution from #11976 did not solve the problem with sorting tests with multiple parameters

Tests

```python
import pytest


@pytest.mark.parametrize('proto', ['serial', 'telnet', 'ssh'], scope='class')
@pytest.mark.parametrize('unit', [1, 2, 3], scope='class')
class TestA:
    def test_one(self, proto, unit):
        pass

    def test_two(self, proto, unit):
        pass
```

Collecting items with pytest 7.4.4

```
11:07 $ pytest --co
=========================================================================== test session starts ===========================================================================
platform linux -- Python 3.11.6, pytest-7.4.4, pluggy-1.4.0 -- /home/lolik/.cache/pypoetry/virtualenvs/pytest_strain-FQAfhJsh-py3.11/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.11.6', 'Platform': 'Linux-6.7.4-arch1-1-x86_64-with-glibc2.39', 'Packages': {'pytest': '7.4.4', 'pluggy': '1.4.0'}, 'Plugins': {'metadata': '3.1.0'}, 'GIT_BRANCH': 'master'}
rootdir: /home/lolik/Projects/straing/pytest
configfile: pytest.ini
plugins: metadata-3.1.0
collected 18 items                                                                                                                                                        

<Package tests>
  <Module test_asd.py>
    <Class TestA>
      <Function test_one[1-serial]>
      <Function test_two[1-serial]>
      <Function test_one[1-telnet]>
      <Function test_two[1-telnet]>
      <Function test_one[1-ssh]>
      <Function test_two[1-ssh]>
      <Function test_one[2-serial]>
      <Function test_two[2-serial]>
      <Function test_one[2-telnet]>
      <Function test_two[2-telnet]>
      <Function test_one[2-ssh]>
      <Function test_two[2-ssh]>
      <Function test_one[3-serial]>
      <Function test_two[3-serial]>
      <Function test_one[3-telnet]>
      <Function test_two[3-telnet]>
      <Function test_one[3-ssh]>
      <Function test_two[3-ssh]>

======================================================================= 18 tests collected in 0.01s =======================================================================
```

Collecting items with pytest 8.0.1

```
11:08 $ pytest --co
=========================================================================== test session starts ===========================================================================
platform linux -- Python 3.11.6, pytest-8.0.1, pluggy-1.4.0 -- /home/lolik/.cache/pypoetry/virtualenvs/pytest_strain-FQAfhJsh-py3.11/bin/python
cachedir: .pytest_cache
metadata: {'Python': '3.11.6', 'Platform': 'Linux-6.7.4-arch1-1-x86_64-with-glibc2.39', 'Packages': {'pytest': '8.0.1', 'pluggy': '1.4.0'}, 'Plugins': {'metadata': '3.1.0'}, 'GIT_BRANCH': 'master'}
rootdir: /home/lolik/Projects/straing/pytest
configfile: pytest.ini
plugins: metadata-3.1.0
collected 18 items                                                                                                                                                        

<Dir pytest>
  <Package tests>
    <Module test_asd.py>
      <Class TestA>
        <Function test_one[1-serial]>
        <Function test_two[1-serial]>
        <Function test_one[2-serial]>
        <Function test_two[2-serial]>
        <Function test_one[2-telnet]>
        <Function test_two[2-telnet]>
        <Function test_one[1-telnet]>
        <Function test_two[1-telnet]>
        <Function test_one[3-telnet]>
        <Function test_two[3-telnet]>
        <Function test_one[3-serial]>
        <Function test_two[3-serial]>
        <Function test_one[3-ssh]>
        <Function test_two[3-ssh]>
        <Function test_one[2-ssh]>
        <Function test_two[2-ssh]>
        <Function test_one[1-ssh]>
        <Function test_two[1-ssh]>

======================================================================= 18 tests collected in 0.01s =======================================================================

```

### PR DESCRIPTION
**Pull request by sadra-barikbin**
In #11220, an unintended change in reordering was introduced by changing the way indices were assigned to direct params. This PR reverts that change and reduces #11220 changes to just refactors.

After this PR we could safely decide on the solutions discussed in #12008, i.e. #12082 or the one initially introduced in #11220 .

Fixes #12008 

@bluetech @RonnyPfannschmidt

### CODE REVIEW

**Review by RonnyPfannschmidt — APPROVED** [review]
thanks for addressing this

**Comment by RonnyPfannschmidt** [comment]
@bluetech please give it a quick check as well, else i'd like to get this in soon

@sadra-barikbin do we need to backport this?

**Comment by sadra-barikbin** [comment]
> @bluetech please give it a quick check as well, else i'd like to get this in soon
> 
> @sadra-barikbin do we need to backport this?

To 8.2.2 and its earlier versions back to 8.0.1 or just 8.2.2?

**Comment by The-Compiler** [comment]
We usually only do patch releases for the latest stable branch, i.e. 8.2.

File: changelog/12008.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+In PR#11220, an unintended change in reordering was introduced by changing the way indices were assigned to direct params. More specifically, before that change, the indices of direct params to metafunc's callspecs were assigned after all parametrizations took place. Now, that change is reverted.
```
Comment (nicoddemus) [inline changelog/12008.bugfix.rst:1]:
```suggestion
In :pr:`11220`, an unintended change in reordering was introduced by changing the way indices were assigned to direct params. More specifically, before that change, the indices of direct params to metafunc's callspecs were assigned after all parametrizations took place. Now, that change is reverted.
```

**Comment by nicoddemus** [comment]
It would be nice to get @bluetech's review here too before we merge this.

**Comment by RonnyPfannschmidt** [comment]
in case its not possible to get ran's feedback this by Sunday, lets merge, i believe its good to go and we shouldn't block for too long just because hes busy

File: src/_pytest/python.py [hunk 2]
```diff
@@ -1445,6 +1449,12 @@ def _validate_if_using_arg_names(
                         pytrace=False,
                     )
 
+    def _recompute_direct_params_indices(self):
```
Comment (nicoddemus) [inline src/_pytest/python.py:1452]:
```suggestion
    def _recompute_direct_params_indices(self) -> None:
```

**Comment by RonnyPfannschmidt** [comment]
sorry for the delay and thanks for the patience

**Comment by RonnyPfannschmidt** [comment]
As bluetech removed the back port label I believe we ought not, else he can revise

**Comment by nicoddemus** [comment]
Ahh I missed that, sorry. We should not backport then. 👍

**Pull request merged**